File: app/src/main/java/adapter/CastAdapter.java
package adapter;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

import br.com.icaro.filme.R;
import domain.CastItem;
import pessoa.activity.PersonActivity;
import utils.Constantes;
import utils.UtilsApp;

/**
 * Created by icaro on 22/02/17.
 */
public class CastAdapter extends RecyclerView.Adapter<CastAdapter.CastViewHolder> {
    private Context context;
    private List<CastItem> casts;

    public CastAdapter(FragmentActivity activity, List<CastItem> cast) {
        context = activity;
        this.casts = cast;
    }


    @Override
    public CastViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.scroll_elenco, parent, false);
        return new CastViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final CastViewHolder holder, final int position) {
        holder.progressBar.setVisibility(View.VISIBLE);
        final CastItem personCast = casts.get(position);
        if (personCast.getName() != null || personCast.getCharacter() != null) {
            holder.textCastPersonagem.setText(personCast.getCharacter());
            holder.textCastNome.setText(personCast.getName());
            
            Picasso.get()
                    .load(UtilsApp.INSTANCE.getBaseUrlImagem(UtilsApp.INSTANCE.getTamanhoDaImagem(context, 4)) + personCast.getProfilePath())
                    .placeholder(R.drawable.person)
                    .into(holder.imageView, new Callback() {
                        @Override
                        public void onSuccess() {
                            holder.progressBar.setVisibility(View.INVISIBLE);
                        }

                        @Override
                        public void onError(Exception e) {
                            holder.progressBar.setVisibility(View.INVISIBLE);
                        }
                    });

        } else {
            holder.textCastPersonagem.setVisibility(View.GONE);
            holder.imageView.setVisibility(View.GONE);
            holder.textCastNome.setVisibility(View.GONE);
            holder.textCastPersonagem.setVisibility(View.GONE);
            holder.progressBar.setVisibility(View.GONE);
        }

        holder.imageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(context, PersonActivity.class);
                intent.putExtra(Constantes.PERSON_ID, personCast.getId());
                intent.putExtra(Constantes.NOME_PERSON, personCast.getName());
                context.startActivity(intent);

                Bundle bundle = new Bundle();
                bundle.putString(FirebaseAnalytics.Event.SELECT_CONTENT, PersonActivity.class.getName());
                bundle.putInt(FirebaseAnalytics.Param.ITEM_ID, personCast.getId());
                bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, personCast.getName());
                FirebaseAnalytics.getInstance(context).logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
            }
        });
    }

    @Override
    public int getItemCount() {
        return casts.size();
    }

    class CastViewHolder extends RecyclerView.ViewHolder {
       private ProgressBar progressBar;
        private ImageView imageView;
        private TextView textCastPersonagem, textCastNome;

        CastViewHolder(View itemView) {
            super(itemView);
            textCastNome = itemView.findViewById(R.id.textCastNomes);
            textCastPersonagem = itemView.findViewById(R.id.textCastPersonagem);
            imageView = itemView.findViewById(R.id.imgPager);
            progressBar = itemView.findViewById(R.id.progressBarCast);
        }
    }
}

File: app/src/main/java/adapter/CastAdapter.kt
package adapter

import adapter.CastAdapter.CastViewHolder
import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import br.com.icaro.filme.R
import br.com.icaro.filme.R.drawable
import br.com.icaro.filme.R.id
import br.com.icaro.filme.R.layout
import domain.CastItem
import pessoa.activity.PersonActivity
import utils.Constant
import utils.gone
import utils.setPicassoWithCacheAndHolder
import utils.visible

/**
 * Created by icaro on 22/02/17.
 */
class CastAdapter(val activity: FragmentActivity, val casts: List<CastItem>) : Adapter<CastViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CastViewHolder = CastViewHolder(parent)

    override fun onBindViewHolder(holder: CastViewHolder, position: Int) {
        holder.bind(casts[position])
    }

    override fun getItemCount(): Int {
        return casts.filter {
            (!it.name.isNullOrBlank() || it.character.isNullOrEmpty())
        }.size
    }

    inner class CastViewHolder(parent: ViewGroup) :
        ViewHolder(LayoutInflater.from(activity).inflate(layout.scroll_elenco, parent, false)) {

        private val progressBar: ProgressBar = itemView.findViewById(id.progressBarCast)
        private val img: ImageView = itemView.findViewById(id.imgPager)
        private val textCastPersonagem: TextView = itemView.findViewById(id.textCastPersonagem)
        private val textCastNome: TextView = itemView.findViewById(id.textCastNomes)

        fun bind(person: CastItem) = with(itemView) {
            progressBar.visible()
            val (_, character, _, _, name, profilePath, id) = person
            textCastPersonagem.text = character
            textCastNome.text = name
            img.setPicassoWithCacheAndHolder(stillPath = profilePath,
                patten = 4,
                img_erro = drawable.person,
                error = { progressBar.gone() },
                sucesso = { progressBar.gone() },
                holder = R.drawable.person
            )
            setOnClickListener {
                context.startActivity(Intent(context, PersonActivity::class.java).apply {
                    putExtra(Constant.NOME_PERSON, name)
                    putExtra(Constant.PERSON_ID, id)
                })
            }
        }
    }
}